### ISSUE TITLE
Error in latest Discord and Discord.Net with a previously working bot

### ISSUE THREAD

**Issue by mcmonkey4eva** [RogueException/Discord.Net#442]
Discord updated recently, and my working bot doesn't work any more, with a hell of a stack trace...
```
2016/12/25 06:12:28 [INFO] System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.Net.Sockets.SocketException: An address incompatible with the requested protocol was used
  at System.Net.Sockets.Socket..ctor (System.Net.Sockets.AddressFamily addressFamily, System.Net.Sockets.SocketType socketType, System.Net.Sockets.ProtocolType protocolType) [0x0005d] in <07e58fee265744b6a3510300e59793c1>:0 
  at System.Net.Sockets.Socket.ConnectAsync (System.Net.Sockets.SocketType socketType, System.Net.Sockets.ProtocolType protocolType, System.Net.Sockets.SocketAsyncEventArgs e) [0x0000b] in <07e58fee265744b6a3510300e59793c1>:0 
  at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
   --- End of inner exception stack trace ---
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00053] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Reflection.MethodBase.Invoke (System.Object obj, System.Object[] parameters) [0x00000] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at SuperSocket.ClientEngine.ConnectAsyncExtension.ConnectAsync (System.Net.EndPoint remoteEndPoint, SuperSocket.ClientEngine.ConnectedCallback callback, System.Object state) [0x00039] in <4ba236fbf29248a4985a92e6e3907e4a>:0 
  at SuperSocket.ClientEngine.TcpClientSession.Connect () [0x00074] in <4ba236fbf29248a4985a92e6e3907e4a>:0 
  at WebSocket4Net.WebSocket.Open () [0x0003a] in <4ba236fbf29248a4985a92e6e3907e4a>:0 
  at Discord.Net.WebSockets.WS4NetEngine.Connect (System.String host, System.Threading.CancellationToken cancelToken) [0x000d8] in <37261fa8187547a4a9befeeb238a1dc3>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at Discord.Net.WebSockets.WebSocket+<BeginConnect>d__34.MoveNext () [0x00420] in <37261fa8187547a4a9befeeb238a1dc3>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0004e] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at Discord.Net.WebSockets.GatewaySocket+<Connect>d__12.MoveNext () [0x00181] in <37261fa8187547a4a9befeeb238a1dc3>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0004e] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at Discord.DiscordClient+<BeginConnect>d__78.MoveNext () [0x003b9] in <37261fa8187547a4a9befeeb238a1dc3>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at Discord.DiscordClient+<BeginConnect>d__78.MoveNext () [0x00653] in <37261fa8187547a4a9befeeb238a1dc3>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0004e] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.ConfiguredTaskAwaitable+ConfiguredTaskAwaiter.GetResult () [0x00000] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at Discord.DiscordClient+<Connect>d__77.MoveNext () [0x0009a] in <37261fa8187547a4a9befeeb238a1dc3>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0004e] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at DenizenIRCBot.dDiscordBot+<>c__DisplayClass2_0+<<Init>b__0>d.MoveNext () [0x00076] in <a248df1f3a6c4908975943335271d989>:0 
--- End of stack trace from previous location where exception was thrown ---
  at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x0000c] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (System.Threading.Tasks.Task task) [0x0004e] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (System.Threading.Tasks.Task task) [0x0002e] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd (System.Threading.Tasks.Task task) [0x0000b] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at System.Runtime.CompilerServices.TaskAwaiter.GetResult () [0x00000] in <dca3b561b8ad4f9fb10141d81b39ff45>:0 
  at Discord.DiscordClient.ExecuteAndWait (System.Func`1[TResult] asyncAction) [0x0000d] in <37261fa8187547a4a9befeeb238a1dc3>:0 
  at DenizenIRCBot.dDiscordBot.Init (DenizenIRCBot.YAMLConfiguration conf) [0x00062] in <a248df1f3a6c4908975943335271d989>:0 
```

**Comment by Emzi0767** [comment]
It's not Discord.Net that updated, it's your installation of Mono. You need to downgrade it to 4.4.2 at most.

**Comment by mcmonkey4eva** [comment]
Oh did my server decide to update itself again? Wonderful.
Why /downgrade/ mono? Why isn't Discord.Net compatible with latest mono?

**Comment by Emzi0767** [comment]
It's a Mono problem. Lots of things are broken with it.

**Comment by mcmonkey4eva** [comment]
Could you direct me to information about this bug in mono, EG the bug report on that end? I'm quite interested in seeing what exactly went wrong.

**Comment by mcmonkey4eva** [comment]
Oh apparently there's an issue for this already, https://github.com/RogueException/Discord.Net/issues/297
Sorry for the annoyance. Will work from that issue!

**Comment by khionu** [comment]
@mcmonkey4eva if your server is autoupdating, you should disable whatever is doing that. That is not default behavior for any Linux distribution I know of, plus it's extremely risky behavior for production environments.

For Ubuntu, if you run `apt-get upgrade`, without care, it will upgrade all your packages, including Mono.

**Comment by mcmonkey4eva** [comment]
@khionu it's neither autoupdated nor a general upgrade.
Looking back at what happened a bit more clearly: I ran a regular install and it said it was going to update mono while installing the other package - hell if I know why - maybe that package relied on a higher mono, I don't entirely remember which bit was installed. I just allowed that on the assumption that it's perfectly fine for mono to update... which it really should be.

I am, of course, not doing this on a large scale production environment, it's a relatively "small" setup that just happens to run a Discord/IRC multibot that I would prefer continue functioning on both sides, rather than IRC only.
